exec/cnex/lib/io.c

```diff
@@ -0,0 +1,189 @@
+#include "io.h"
+#include "enums.h"
+
+#ifdef _MSC_VER
+#define _CRT_SECURE_NO_WARNINGS
+#endif
+#ifdef _WIN32
+#include <io.h>
+#endif
+#include <assert.h>
+#include <iso646.h>
+#include <stdio.h>
+#include <string.h>
+#ifndef _WIN32
+#include <unistd.h>
+#include <sys/types.h>
+#endif
+
+#include "cell.h"
+#include "exec.h"
+#include "number.h"
+#include "stack.h"
+#include "nstring.h"
+
+
+static FILE *check_file(TExecutor *exec, void *pf)
+{
+    FILE *f = (FILE *)(pf);
+    if (f == NULL) {
+        exec->rtl_raise(exec, "InvalidFile", "", BID_ZERO);
+        // Exception code will return from here, since f is already NULL.
+    }
+    return f;
+}
+
+
+void io_fprint(TExecutor *exec)
+{
+    Cell *s = peek(exec->stack, 0);
+    void *pf = peek(exec->stack, 1)->address;
+
+    FILE *f = check_file(exec, pf);
+    fwrite(s->string->data, 1, s->string->length, f);
+    fputs("\n", f);
+
+    pop(exec->stack);
+    pop(exec->stack);
+}
+
+void io_close(TExecutor *exec)
+{
+    void *ppf = peek(exec->stack, 0)->address->address;
+
+    FILE *f = check_file(exec, ppf);
+    fclose(f);
+
+    pop(exec->stack);
+}
+
+void io_open(TExecutor *exec)
+{
+    Number mode = top(exec->stack)->number; pop(exec->stack);
+    char *pszName = string_asCString(top(exec->stack)->string); pop(exec->stack);
+
+    const char *m;
+    switch (number_to_uint32(mode)) {
+        case ENUM_Mode_read:  m = "r"; break;
+        case ENUM_Mode_write: m = "w+"; break;
+        default:
+            free(pszName);
+            push(exec->stack, cell_fromAddress(NULL));
+            return;
+    }
+
+    Cell *r = cell_newCellType(cAddress);
+    r->address = (void*)fopen(pszName, m);
+    free(pszName);
+
+    push(exec->stack, r);
+}
+
+void io_readBytes(TExecutor *exec)
+{
+    Number count = top(exec->stack)->number; pop(exec->stack);
+    void *pf = top(exec->stack)->address; pop(exec->stack);
+
+    FILE *f = check_file(exec, pf);
+    uint64_t ncount = number_to_uint64(count);
+
+    Cell *r = cell_createStringCell(ncount);
+    size_t n = fread((unsigned char *)r->string->data, 1, ncount, f);
+
+    assert(r->string->length == n);
```

Comment:
This assert is going to fail the first time you read fewer bytes than you ask for. You'll have to truncate the `r` string to match the number of bytes actually returned by `fread`.